### PR TITLE
Remove with query

### PR DESCRIPTION
**Pull request by wmathurin**
Native code change and tests
The changes for hybrid will come in a separate pull request

### CODE REVIEW

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 1]
```diff
@@ -341,7 +344,7 @@ -(void) testRangeQueryWithStringIndex
  */
 -(void) testRangeQueryWithJSON1Index
 {
-    [self tryAllQuery:kSoupIndexTypeJSON1];
+    [self tryRangeQuery:kSoupIndexTypeJSON1];
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: Unrelated to remove by query - bad copy/paste job when adding tests for json1

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 2]
```diff
@@ -391,15 +396,15 @@ -(void) tryRangeQuery:(NSString*)indexType
  */
 -(void) testLikeQueryWithStringIndex
 {
-    [self tryRangeQuery:kSoupIndexTypeString];
+    [self tryLikeQuery:kSoupIndexTypeString];
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: Unrelated to remove by query - bad copy/paste job when adding tests for json1

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 3]
```diff
@@ -391,15 +396,15 @@ -(void) tryRangeQuery:(NSString*)indexType
...
 }
 
 /**
  * Test like query when soup has json1 index
  */
 -(void) testLikeQueryWithJSON1Index
 {
-    [self tryAllQuery:kSoupIndexTypeJSON1];
+    [self tryLikeQuery:kSoupIndexTypeJSON1];
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: Unrelated to remove by query - bad copy/paste job when adding tests for json1

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 4]
```diff
@@ -363,23 +366,25 @@ -(void) tryRangeQuery:(NSString*)indexType
         XCTAssertNil(error, @"There should be no errors.");
         
         // Populate soup
-        NSDictionary* soupElt1 = @{@"key": @"ka1", @"value":@"va1", @"otherValue":@"ova1"};
-        NSDictionary* soupElt2 = @{@"key": @"ka2", @"value":@"va2", @"otherValue":@"ova2"};
-        NSDictionary* soupElt3 = @{@"key": @"ka3", @"value":@"va3", @"otherValue":@"ova3"};
+        NSDictionary* soupElt0 = @{@"key": @"ka1", @"value":@"va1", @"otherValue":@"ova1"};
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: Test easier to read that way we do soupEltsCreated[x] below

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 5]
```diff
@@ -421,64 +426,248 @@ -(void) tryLikeQuery:(NSString*)indexType
         XCTAssertNil(error, @"There should be no errors.");
         
         // Populate soup
-        NSDictionary* soupElt1 = @{@"key": @"abcd", @"value":@"va1", @"otherValue":@"ova1"};
-        NSDictionary* soupElt2 = @{@"key": @"bbcd", @"value":@"va2", @"otherValue":@"ova2"};
-        NSDictionary* soupElt3 = @{@"key": @"abcc", @"value":@"va3", @"otherValue":@"ova3"};
-        NSDictionary* soupElt4 = @{@"key": @"defg", @"value":@"va1", @"otherValue":@"ova1"};
+        NSDictionary* soupElt0 = @{@"key": @"abcd", @"value":@"va1", @"otherValue":@"ova1"};
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: Test easier to read that way we do soupEltsCreated[x] below

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 6]
```diff
@@ -421,64 +426,248 @@ -(void) tryLikeQuery:(NSString*)indexType
...
+                                        covering:NO
+                             expectedDbOperation:@"SCAN"
                                            store:store];
     }
 }
 
 /**
+ * Test smart query when soup has string index
+ */
+-(void) testSmartQueryWithStringIndex
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: Unrelated to remove by query - new tests for json1 vs string

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 7]
```diff
@@ -421,64 +426,248 @@ -(void) tryLikeQuery:(NSString*)indexType
...
+                                        covering:![indexType isEqualToString:kSoupIndexTypeJSON1] // interestingly the explain plan doesn't use a covering index with a functional index
+                             expectedDbOperation:@"SCAN"
+                                           store:store];
+    }
+}
+
+/**
+ * Test remove entries with ids
+ */
+-(void) testRemoveEntriesByIds
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: There was no (native) test for remove entries by id

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 8]
```diff
@@ -421,64 +426,248 @@ -(void) tryLikeQuery:(NSString*)indexType
...
+                                 expectedResults:@[soupEltsCreated[2]]
+                                        covering:NO
+                             expectedDbOperation:@"SCAN"
+                                           store:store];
+    
+    }
+}
+
+/**
+ * Test remove entries by query
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: Test for the new api

File: libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m [hunk 9]
```diff
@@ -512,19 +701,20 @@ - (void) tryAggregateQueryOnIndexedField:(NSString*) indexType
...
                                             page:0
                                  expectedResults:@[@[[NSNumber numberWithDouble:20.1]]]
+                                        covering:NO
                              expectedDbOperation:nil
                                            store:store];
     }
 }
 
--(void) runQueryCheckResultsAndExplainPlan:(SFQuerySpec*)querySpec page:(NSUInteger)page expectedResults:(NSArray*)expectedResults expectedDbOperation:(NSString*)expectedDbOperation store:(SFSmartStore*)store
+-(void) runQueryCheckResultsAndExplainPlan:(SFQuerySpec*)querySpec page:(NSUInteger)page expectedResults:(NSArray*)expectedResults covering:(BOOL)covering expectedDbOperation:(NSString*)expectedDbOperation store:(SFSmartStore*)store
```
Comment (wmathurin) [inline libs/SmartStore/SmartStoreTests/SFSmartStoreTests.m]: More precise check of explain plan

**Comment by bhariharan** [comment]
👍

**Pull request merged**